# Bevezető

Egy program során különöző számokat és szövegeket manipulálunk, hogy megkapjuk a kívánt végeredményt (például két szám összegét). És ezek az értékek (számok, szövegek) különbözőképpen viselkednek, mert különböző típusúak.

# Számok

A szám típusú értékeket, mint például a `4` vagy `2.0` már láttuk párszor. Össze lehet őket adni, ki lehet őket íratni stb. De nézzük meg egészen pontosan, hogy hogyan is működnek. A programozásban még a számokat is további kategóriákra szokás bontani. Pythonban egészen pontosan kettőre: az egész számok és a tört számok.

## Egész számok (integer)

Egész számokat összeadhatunk, kivonhatunk és szorozhatunk bármilyen probléma nélkül (sőt hatványozhatunk is). Néhány példa:

In [5]:
print(2+1)
print(5-2)
print(3*4)

3
3
12


A `type` függvénynek egy paramétere van és visszatér a paraméterben kapott érték típusúval. Itt például minden esetben `int` típusú értékeink vannak. Az int az integer rövidítése, ami az egész számokat jelenti.

In [7]:
print(type(2))
print(type(2+1))
print(type(5-2))
print(type(3*4))

<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>


Osztásnál arra kell figyelni, hogy az eredmény már nem egész típusú lesz. Ezt jelzi az is, hogy nem `3` hanem `3.0`-ként jelenik meg az eredmény.

In [8]:
print(6/2)
print(type(6/2))

3.0
<class 'float'>


## Tört számok (float)

A tizedes tört számokra lebegőpontos számként szoktunk hivatkozni, innen ered az angol `float` kifejezés. Mégnéztük, hogy az alapvető művelete működnek egész és egész szám között. Megfigyelhetjük, hogy ez igaz tört és tört számokra is sőt, várható, hogy a matematikai műveletek tört és egész számok között is értelmezve vannak.

In [9]:
x = 2 + 2.2
print(x)
print(type(x))

x = 2.6 - 2
print(x)
print(type(x))

x = 2.1 * 3
print(x)
print(type(x))

x = 2/0.5
print(x)
print(type(x))

4.2
<class 'float'>
0.6000000000000001
<class 'float'>
6.300000000000001
<class 'float'>
4.0
<class 'float'>


Megfigyelhetjük azt is, hogy a `2.6 - 2` eredménye nem pontosan `0.6` lett, hanem `0.6000000000000001`. Ennek az az oka, hogy a számítógépben tört számokat nem tudunk teljesen pontosan tárolni.

# Szövegek (string)

Egy másik nagyon fontos ítpus, ami már előkerült azok a szövegek. A szövegeket mindig idézőjel `"` vagy aposztóf jel `'` közé kell tennünk. Amit nem jelölönk így, azt változó vagy esetleg függvény név ként próbálja meg értelmezni a számítógép és hiba üzenetet kapunk ha nem létezik olyan.

In [11]:
print(Az x az ábécé utolsó betűi közé tartozik) # itt minden változó

SyntaxError: invalid syntax (798644086.py, line 1)

In [14]:
print("Az x az ábécé utolsó betűi közé tartozik") # itt ez egy szöveg
print("Az", "x", "az", "ábécé", "utolsó", "betűi", "közé", "tartozik") # itt 8 szöveg (string) típusú paraméter van

Az x az ábécé utolsó betűi közé tartozik
Az x az ábécé utolsó betűi közé tartozik


In [13]:
x = 7
print("Az", x, "az ábécé utolsó betűi közé tartozik") # itt az x egy változó

Az 7 az ábécé utolsó betűi közé tartozik


In [15]:
print(type("szöveg"))

<class 'str'>


A szöveg típust pythonban `str`-nek hívjuk (a string rövidítése).

Szövegek között már nem annyira egyértelmű, hogy létezik-e például az összeadás vagy a kivonás művelete. Tippelj melyik az amelyik értelmezve van és vajon mi történik?

In [17]:
print("alma" + "fa")

szo1 = "körte"
szo2 = "villany"

print(szo2 + szo1)

almafa
villanykörte


A szövegek összeadása tehát működik és a két string összefűzését jelenti. Ezzel szemben szövegeket nem tudunk egymásból kivonni, vagy összeszorozni vagy osztani. Ezek értelmetlenek pythonban és hiba üzenetet kapunk.

In [18]:
print("alma" - "ma") # Hiba!

TypeError: unsupported operand type(s) for -: 'str' and 'str'

In [19]:
print("alma" * "körte") # Hiba!

TypeError: can't multiply sequence by non-int of type 'str'

In [20]:
print("alma" / "alma") # Hiba!

TypeError: unsupported operand type(s) for /: 'str' and 'str'

Vajon szövegek és számok között milyen műveletek vannak értelmezve? Szövegeket nem tudunk összeadni számokkal, még akkor sem, ha az idézőjel között valamilyen szám van.

In [21]:
print("6" + 3)

TypeError: can only concatenate str (not "int") to str

Ugyanígy osztani és kivonni sem tudunk. Azonban a szorzás értelmezve van. Ilyenkor annyiszor ismértli meg és fűzi a végére, amennyivel megszoroztuk

In [23]:
print(3*"ha")
print("ha"*3)

hahaha
hahaha


Sőt, mivel a szorzás hamarabb kerül végrehajtásra a műveleti sorrend miatt, így ilyet is írhatunk akár:

In [25]:
print("Mu" + 3*"ha")
print("Mu" + (3*"ha"))

Muhahaha
Muhahaha


Itt figyelni kell rá, hogy a szorzás is csak egész számok (int) és szövegek között van értelmezve.

In [26]:
print("ha" * 1.5)

TypeError: can't multiply sequence by non-int of type 'float'

In [27]:
print("ha" * 1.0)

TypeError: can't multiply sequence by non-int of type 'float'

Tehát ha egy osztás eredményével szeretnénk szöveget szorozni, az hibás lesz, mert az osztás eredménye minden esetben tört szám.

In [28]:
x = 4/2
print(x*"Alma")

TypeError: can't multiply sequence by non-int of type 'float'

>Figyelem:\
Ha számokat akarunk összeadni figyelni kell rá, hogy nem szövegként vannak-e eltárolva, különben érdekes eredményeket kaphatunk. Ilyenkor ugyanis csak egymás mellé fűzi a két szöveget, ahelyett, hogy összeadná a számokat.

In [29]:
x = "1"
y = "1"
print(x+y)

11


Ugyanez történik akkor is, ha az `input` segítségével számot próbálunk bekérni a felhasználótól. Ugyanis, az `input` minden esetben szöveget ad vissza, még ha az a szöveg úgy is néz ki hogy `"123"`.

In [31]:
szám1 = input("Adjon meg egy számot: ")
szám2 = input("Adjon meg még egy számot: ")

print("A két szám összege:", szám1 + szám2)

Adjon meg egy számot: 2
Adjon meg még egy számot: 4
A két szám összege: 24


# Átjárás típusok között (type casting)

Mit tehetünk a fentebb említett esetekben, ha nekünk mégis számként lenne szükségünk arra a szövegre? 

Létezik 3 függvény (`int`, `float`, `str`), aminek segítségével átalakíthatjuk egyik típust a másikba. De csak akkor ha ez értelmesen megtehető, például az `"alma"` szöveg értelem szerűen nem alakítható át számmá, azonban a `"22"` mint szöveg egyszerűen átalakítható számmá. Ha az átalakítás nem végezhető el, akkor hibát kapunk.

Szöveggé bármilyen egyszerű típus átalakítható (egész, tört számok). Néhány példa:

In [32]:
x = "12" # egy szöveg típusú változó

# print(x + 2)-> ez így hibás lenne, mert x szöveg típusú

print(int(x) + 2)

14


In [34]:
szám = input("Adjon meg egy számot: ")

# print(szám * 2) -> ez így hibás lenne, mert a szám változóban szöveg típusú érték van

szám = int(szám) # a szám változó értékét lecseréljük az átalakított értékre
print(szám,"kétszerese:",szám * 2) # az előző sor miatt ez már helyes

Adjon meg egy számot: 22
22 kétszerese: 44


A függvényeket egymásba is ágyazhatjuk. Ahelyett, hogy különsorban bekérnénk a felhasználótól egy számot, majd átalakítanánk annak a típusát. Ezt leírhatjuk egyben, egy sorban. Bekérünk egy számot, majd a bekért számot átkonvertáljuk (átalakítjuk) rögtön. EZ azért lehetséges, mert az `input` függvény visszatér azzal az értékkel ("a gép kidobja magából azt az értéket"), amit megadott a felhasználó. Ezért tudtuk egy változóban eltárolni is.

In [35]:
szám = int(input("Adjon meg egy számot: "))

print("A megadott szám háromszorosa:", szám*3)

Adjon meg egy számot: 11
A megadott szám háromszorosa: 33


In [37]:
x = "00123"
print(int(x))

123


In [36]:
x = "1.1"
print(int(x)) # Hiba! Egész számmá csak az a szöveg alakítható, ami csak számokat tartalmaz

ValueError: invalid literal for int() with base 10: '1.1'

In [39]:
x = "1.1"

print(x*2)
print(float(x) * 2)

1.11.1
2.2


In [40]:
x = 12

print(x*2) # itt x szám
print(str(x)*2) # itt x szöveg

24
1212
